Add newlines and simplify some parts

weebl2000 opened the pull request:
No functional changes, just increase readability of code and docs and sprinkle some more resilient programming idioms.

Some changes seem very stupid/trivial, but for [example](https://github.com/nfcgate/nfcgate/pull/149/files#diff-a3869c620922aa1408c1015c1d5c0bfb8c5c60f84d68f2ffc523bf92958e8ff5R221):

![image](https://github.com/nfcgate/nfcgate/assets/2385844/b46036de-4fdc-48ee-8514-cd857dfcc7e5)

Having two separate return statements will make it immensively more trivial to debug a method. You can just set your breakpoint on the return statement to see where the logic went, instead of having to backtrack the single return statement.

Of course, ideally, all variables are only ever assigned once! The `result` variable is used in a dirty way here, but I didn't want to refactor too much in one go.

roussosalex commented:
Thanks for your contribution (again³)! Unfortunately, this PR is very hard to read. Please separate the different types of changes more cleanly into separate commits (e.g. the "date formatter" change, adding "final" to variables, changing callbacks to lambdas, etc. could all be separate commits).

weebl2000 commented:
> Thanks for your contribution (again³)! Unfortunately, this PR is very hard to read. Please separate the different types of changes more cleanly into separate commits (e.g. the "date formatter" change, adding "final" to variables, changing callbacks to lambdas, etc. could all be separate commits).

Sure, I'll try to do that, staring with the most important stuff and the nitpick stuff later.

Commit:
Create new SimpleDateFormatter every time

SimpleDateFormatter is not thread-safe, if you use it concurrently it
will mangle the dates. It's best to create a new one every time.

Commit:
Deduplicate .gitignore and add common files that should be ignored

Commit:
Use single quotes in build.gradle for non-variables

It's best to use double quotes only when there's a variable in the
string

Commit:
Make documentation a little more readable

Commit:
Add language tag to html file

Commit:
Make every text file end with a newline

Nitpicky, but POSIX 3.206 Line definition

Commit:
Specify Java version in build.gradle

Use Java 8 for now

Commit:
Make implicity final variables explicity final

Marking them as explicitly final makes it clear to anyone reading the
code it will only be assigned once.

Commit:
Simplify empty xml tag

Commit:
Set interrupt status instea of ignoring the exception

Commit:
Use try-with-resources on OutputStream

It's better to let java handle the closing than doing it yourself

Commit:
Replace case-switch with an if statement

The case-switch is a bit obscure for this use-case. Using an
if-statement more clearly shows the application flow

Commit:
Replace anonymous inner-classes with lambda statements

No logic is changed, but the code becomes more readable with less
boilerplate

Commit:
Remove unused imports

Commit:
Make inner classes static

Since they don't need any variables from the parent class

Commit:
Replace case-switch with if-statement

The IDs will no longer be final

Commit:
Use static method reference instead of creating a new instance of the class

Commit:
More explicitly final variables

Commit:
Remove + 0

These calls are silly, + 0 has no effect

Commit:
Variable is method-scoped, declare it at method level

Commit:
Remove unused variable

Commit:
Change variable to method scope

Commit:
This variable is effectively final, don't need null assignment

Commit:
Prevent warning on always returning false

Since it will be overriden by the native module, prevent IDE warnings

Commit:
Remove redundant null check

We just assigned the variable, it won't be null

Commit:
Supress warning for unregistered receiver flags

Since it's only happening in the old API call

Commit:
Remove throws Throwable definition

It's not very useful to declare throws Throwable

Commit:
Don't use raw class

Commit:
Remove final modifier from static method

Commit:
Use try-with-resources when working with stream

This makes sure it will always be closed, even on exceptions

Commit:
Formatting

weebl2000 commented:
I've split up the changes from 4 commits into 40 commits which will make it easier to review all changes on a per-commit basis.

Also: it's probably a good idea to enable the "hide whitespace" option when reviewing this commit. It will reduce the clutter when viewing changes.

![image](https://github.com/nfcgate/nfcgate/assets/2385844/c69d24fb-e175-4d6c-9e8d-9cc9ea8f1268)

roussosalex commented:
Wow! That is a lot of commits, but much more readable now. Thank you for your hard work on this PR.

While most changes seem to look good, I can already see some that will need to be adjusted for consistency with our intended code-style. It may take a little time to review each commit carefully.

roussosalex commented on app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/gui/component/FileShare.java:
```diff
@@ -51,18 +51,15 @@ public FileShare setMimeType(String mimeType) {
-            file = new File(shareDir, mPrefix + mExtension);
+        // create file with given prefix and extension
+        final File file = new File(shareDir, mPrefix + mExtension);
+        try (final OutputStream stream = new FileOutputStream(file)){
```
```suggestion
        try (final OutputStream stream = new FileOutputStream(file)) {
```

weebl2000 replied:
Done.

roussosalex commented on app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/network/data/NetworkStatus.java:
```diff
@@ -2,7 +2,6 @@
 
 public enum NetworkStatus {
     ERROR,
-
```
Please leave this line as is, because it is used to visually separate the error status codes from the operational codes

roussosalex reviewed (changes requested):
In addition to the changes listed below, there are some per-commit comments:

a0fc2d2ddecb1edf1aedaa3dd834e2c297cc9c4e:  
The switch used is not "obscure" but in fact standard Android code, only recently has `R.id` been changed to non-final, therefore requiring if/else instead of switch.

557cbd5b77b40a0ae84634825ab80e62e828a191
1ce14e4b2eafa314e8895926b21e2fee84e4229e:
Please remove Objects.requireNonNull from these commits as it worsens readability and does not add significant benefit. 

cae340b134087bc51583929b5f0eb058ced6cdd7:
Please remove this commit, because the JavaDoc is formatted consistently without HTML tags.

67de060fcefedc08961795b9d019be0f66e549d9:
Please remove this commit, because the null-check was added for a reason.

3cabca8ace6084a31ea9322e971cc858e4134ea7:
Please fix the typo in the commit message

27d1c18f1d9b01bcf31ab97d751dff20328ab32c:
Please remove this commit, because all `detectFoo()` methods are formatted consistently and structured the same way.

6b22b78f899ef40b07acfa32ff75bb5c00359af6:
These variables were intentionally put in the class scope for future extendability, but we can move them to the method-scope for now.

d76218d38abe9d9b033fde9a09eddd65dda75f89:
Please remove this commit, because we intentionally did not use singletonList for future extendability. 

b66b424cf8a029890aa95f59bbb75b12d0567f3c:
Please remove this commit, because it worsens the readability of the code, reduces the performance and is not functionally the same code, while not providing any tangible benefits over the existing solution.

ecd531e057cb7da1550d91a529557e9482d69e10:
Please remove this commit, because it worsens the readability of the code and does not provide any tangible benefit over returning outside of the catch.

ec6c12b0914ee21ebfb74a642b1fc610d3fff1d4:
Please remove this commit, because it strongly reduces the readability of the code. We can safely rely on the "return statement" to do its job and not have to unnecessarily increase the nesting level.

a4aa9ceea029f7fabb444e64cc0262a85b66ab95:
Please remove this commit, because it unnecessarily adds debugging code. There are conditional breakpoints for the use-case described in the commit, the existing code does not need to be changed. 

Thank you for your effort in trying to improve NFCGate.
On a general note, the App developers usually do not need an explanation of Java basics, that seems just a little bit condescending.

weebl2000 committed:
Revert Objects.requireNonNull

weebl2000 committed:
Remove this one too

weebl2000 committed:
Add space

weebl2000 committed:
Add space

weebl2000 commented:
> [a0fc2d2](https://github.com/nfcgate/nfcgate/commit/a0fc2d2ddecb1edf1aedaa3dd834e2c297cc9c4e):
> The switch used is not "obscure" but in fact standard Android code, only recently has `R.id` been changed to non-final, therefore requiring if/else instead of switch.

I mainly meant the case-switch was a bit overkill for the one condition. Case-switch is of course very useful for checking against many values. I've replaced the other case-switches now too: https://github.com/nfcgate/nfcgate/pull/149/commits/b5c76fffa9c1fbbd3fb529cbf834919abb25a8fa

> [67de060](https://github.com/nfcgate/nfcgate/commit/67de060fcefedc08961795b9d019be0f66e549d9):
> Please remove this commit, because the null-check was added for a reason.

It really cannot be null though?

![image](https://github.com/nfcgate/nfcgate/assets/2385844/c69f58cc-7366-400e-bd45-f4066410e900)

> [d76218d](https://github.com/nfcgate/nfcgate/commit/d76218d38abe9d9b033fde9a09eddd65dda75f89):
> Please remove this commit, because we intentionally did not use singletonList for future extendability.

Sure, done, but be careful of YAGNI ;). Can always change it later.

> [b66b424](https://github.com/nfcgate/nfcgate/commit/b66b424cf8a029890aa95f59bbb75b12d0567f3c):
> Please remove this commit, because it worsens the readability of the code, reduces the performance and is not functionally the same code, while not providing any tangible benefits over the existing solution.

Done, and yes it does not do exactly the same thing, but it will do exactly what you want, avoid double slashes. To me the code improves the readability: "format two strings with a slash in between and replace all double occurences of slashes with a single one". I don't think the performance of this is measurably different. ;) With Java NIO it could be rewritten to `Paths.get(p1, p2)`

> [ecd531e](https://github.com/nfcgate/nfcgate/commit/ecd531e057cb7da1550d91a529557e9482d69e10):
> Please remove this commit, because it worsens the readability of the code and does not provide any tangible benefit over returning outside of the catch.

Removed, but I guess this is personal preference as to me it increases readability to have the return as close as possible to where the exception occurs. And it does have an added benefit, you cannot write code under the try-catch block now without moving the return first. So it will force developers to think about what they are doing if they do write code under the block.

> On a general note, the App developers usually do not need an explanation of Java basics, that seems just a little bit condescending.

This was never my intention, which comment are you refering to me? In any case, sorry for that.

weebl2000 committed:
Replace case-switch with conditions since id is non-final

As suggested

roussosalex commented on app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/gui/fragment/CaptureFragment.java:
```diff
@@ -40,18 +40,8 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
-            }
-        });
+        mStartButton.setOnClickListener(v12 -> startCapture());
+        mStopButton.setOnClickListener(v1 -> stopCapture());
```
Please rename these to `view` for consistency

weebl2000 replied:
Good one, it was auto generated by IntelliJ. I dislike obscure parameters too.

roussosalex commented on app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/gui/fragment/CloneFragment.java:
```diff
@@ -65,35 +64,27 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
+            if (position >= 0) {
+                // load configuration of saved tag
+                final TagInfo item = mTagInfoAdapter.getItem(position);
+                assert item != null;
```
Please remove this assertion for consistency with the other handlers

roussosalex commented on app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/gui/MainActivity.java:
```diff
@@ -158,26 +147,19 @@ protected void onPause() {
-                return new LoggingFragment();
-            default:
-                throw new IllegalArgumentException("Position out of range");
+        final Fragment fragment = R.id.nav_clone == id ? new CloneFragment()
```
Please change this obscure ternary chain to a regular if/else block for consistency

roussosalex commented on nfcd/src/main/java/de/tu_darmstadt/seemoo/nfcgate/xposed/Hooks.java:
```diff
@@ -213,16 +212,17 @@ private Object loadOrInjectClass(Context ctx, String sourcePackage,
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
             return injectClass(ctx, sourcePackage, target, className);
-        else
+        } else {
```
Please remove these braces for consistency with the rest of the code

weebl2000 replied:
Okay, will do. I'd be more than happy to add braces to every statement currently missing them too though, for consistency. 😝 

Having braces everywhere is low hanging fruit to reduce the number of bugs in code. :) Compilers are smart enough to compile it to the same byte code, why would you want to risk the humans making the mistake of thinking if they need curly brackets or not? Just always use them! Nearly all code styles use brackets.

Not using braces is the human equivalent of not using your indicator lights in your car "because nobody is there to see them". Instead of having to spend time to figure if someone can see them, just use them.

https://cafe.elharo.com/blogroll/braceless-if-considered-harmful/

roussosalex replied:
No statements are currently "missing" any braces, they have been intentionally left out to improve readability.

> which comment are you refering to me

Here is an example

weebl2000 replied:
> > which comment are you refering to me

Hehe, using braces is objectively better though. But I will remove them 😄.

roussosalex reviewed (changes requested):
Thanks for improving the PR so fast. 

In addition to the changes requested below, there is still 67de060fcefedc08961795b9d019be0f66e549d9 to remove.

Please also consolidate the commits in this PR ("resolve" the reverts) so we can merge them by rebase and keep your commit messages.

weebl2000 commented:
> Thanks for improving the PR so fast.
> 
> In addition to the changes requested below, there is still [67de060](https://github.com/nfcgate/nfcgate/commit/67de060fcefedc08961795b9d019be0f66e549d9) to remove.
> 
> Please also consolidate the commits in this PR ("resolve" the reverts) so we can merge them by rebase and keep your commit messages.

It might be easier to use github's squash commit option. If not I can do redo the commits.

roussosalex commented:
> I mainly meant the case-switch was a bit overkill for the one condition

It was not overkill, it was auto-generated code and widely used at the time to provide extensibility.

> Sure, done, but be careful of YAGNI

I do not think that applies here, given that making the list non-extensible is more effort than leaving it extensible. 

> Done, and yes it does not do exactly the same thing, but it will do exactly what you want, avoid double slashes

Except that we do not want to avoid all double slashes, only ones at the specific join spot. The string-replace method already fails at the simple case of `combinePath("/a/b/c/", "/d/e/f") // yields "/a/b/c//d/e/f"`.

> It might be easier to use github's squash commit option

The point of splitting this PR into commits was not just to ease the review process, but to properly document the reason behind appropriately grouped changes, instead of having one massive commit.

weebl2000 committed:
Implement PR remarks

weebl2000 commented:
> > I mainly meant the case-switch was a bit overkill for the one condition
> 
> It was not overkill, it was auto-generated code and widely used at the time to provide extensibility.
> 
> > Sure, done, but be careful of YAGNI
> 
> I do not think that applies here, given that making the list non-extensible is more effort than leaving it extensible.
> 
> > Done, and yes it does not do exactly the same thing, but it will do exactly what you want, avoid double slashes
> 
> Except that we do not want to avoid all double slashes, only ones at the specific join spot. The string-replace method already fails at the simple case of `combinePath("/a/b/c/", "/d/e/f") // yields "/a/b/c//d/e/f"`.
> 
> > It might be easier to use github's squash commit option
> 
> The point of splitting this PR into commits was not just to ease the review process, but to properly document the reason behind appropriately grouped changes, instead of having one massive commit.

I'm a huge fan of having the git history reflect exactly every change. I will clean it up this weekend. I was just asking because I think my last PR was squashed. If you're actually rebasing it makes sense for me to split it up.

roussosalex commented:
Sometimes, it makes sense to squash, especially with smaller changes. In this case, I would prefer to rebase.

weebl2000 commented:
Cleaned up the history and implemented the last remarks with mostly rebases. I checked as best as I could and think I didn't miss anything, but this was a little Git sorcercy. Let me know if I missed anything.

roussosalex commented:
Thanks for delving into git sorcery for this PR!

There are still the commits 4413d82f1c2885801f56fb3c90cc27f878e5c228 and 047989d1b19626f9a3fc42c1608b006a3b62be82 to remove, as discussed above.

weebl2000 commented:
> Thanks for delving into git sorcery for this PR!
> 
> There are still the commits [4413d82](https://github.com/nfcgate/nfcgate/commit/4413d82f1c2885801f56fb3c90cc27f878e5c228) and [047989d](https://github.com/nfcgate/nfcgate/commit/047989d1b19626f9a3fc42c1608b006a3b62be82) to remove, as discussed above.

Dropped https://github.com/nfcgate/nfcgate/commit/047989d1b19626f9a3fc42c1608b006a3b62be82

I'm more than happy to drop https://github.com/nfcgate/nfcgate/commit/4413d82f1c2885801f56fb3c90cc27f878e5c228 too, but what am I missing here? The variable is declared on class level as final. https://github.com/nfcgate/nfcgate/blob/4413d82f1c2885801f56fb3c90cc27f878e5c228/app/src/main/java/de/tu_darmstadt/seemoo/nfcgate/gui/fragment/ReplayFragment.java#L37

It should really never ever be able to become null? Unless there's some obscure Java/Android thingy going on? If that's the case maybe I should add a comment as to why the null check is there?

roussosalex commented:
> It should really never ever be able to become null

Yes, that is right. I don't remember the exact scenario, but there used to be an issue there with stop/resume that we fixed (by introducing the null-check) a few years ago. It is possible that this issue can no longer occur with the current code. In the end, keeping the check for now does not hurt. We may re-evaluate the necessity of the check in the future
